packages/core/src/storage/extract.ts

```diff
@@ -17,62 +22,98 @@ export function extractStorageLayout(
   contractDef: ContractDefinition,
   decodeSrc: SrcDecoder,
   deref: ASTDereferencer,
+  storageLayout?: StorageLayout | undefined,
 ): StorageLayout {
-  const layout: StorageLayout = { storage: [], types: {}, layoutVersion: currentLayoutVersion };
+  const layout: StorageLayout = { storage: [], types: {}, layoutVersion: currentLayoutVersion, flat: false };
+  if (storageLayout !== undefined) {
+    layout.types = storageLayout.types;
+    if (layout.types) {
+      for (const type of Object.entries(layout.types)) {
+        if (type[1].members) {
+          for (const member of type[1].members) {
+            // If member includes contract and ast the declaration contract should be used, right now it always uses the contract being processed
+            if (typeof member === 'object' && 'contract' in member && member.contract) {
+              const [, contract] = getOriginContract(contractDef, member.astId, deref);
+              member.contract = contract;
+            }
+          }
+        } else {
+          // In some cases the past version didnt have members and the updated version has members = undefined,
+          // this is for those cases
+          type[1].members = undefined;
+        }
+      }
+    }
 
-  // Note: A UserDefinedTypeName can also refer to a ContractDefinition but we won't care about those.
-  const derefUserDefinedType = deref(['StructDefinition', 'EnumDefinition']);
+    for (const storage of storageLayout.storage) {
+      let varDecl: any = contractDef.nodes.filter(n => n.id == storage.astId)[0];
```

Comment:
No need for this optimization IMO. It should be handled in `getOriginContract`, instead of skipping the "base contract" in that function.